universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java

```diff
@@ -2,55 +2,98 @@
 import java.time.Instant;
+import java.util.HashMap;
 import java.util.Locale;
 import java.util.Objects;
 import java.util.Optional;
 import javax.annotation.Nullable;
 
 public class ApplicantData {
+  // Suppress errors thrown by JsonPath and instead return null if a path does not exist in a JSON
+  // blob.
+  private static final Configuration CONFIGURATION =
+      Configuration.defaultConfiguration().addOptions(Option.SUPPRESS_EXCEPTIONS);
 
   private DocumentContext jsonData;
   private static final String EMPTY_APPLICANT_DATA_JSON = "{ \"applicant\": {}, \"metadata\": {} }";
 
   public ApplicantData() {
-    this(JsonPath.parse(EMPTY_APPLICANT_DATA_JSON));
+    this(EMPTY_APPLICANT_DATA_JSON);
   }
 
-  public ApplicantData(DocumentContext jsonData) {
-    this.jsonData = checkNotNull(jsonData);
+  public ApplicantData(String jsonData) {
+    this.jsonData = JsonPath.using(CONFIGURATION).parse(checkNotNull(jsonData));
   }
 
   public Locale preferredLocale() {
     return Locale.ENGLISH;
   }
 
   /**
-   * Attempts to read a string at the given path in the applicant's answer data. Returns an {@code
-   * Optional.empty()} if the path does not exist.
+   * Puts the given value at the given path in the underlying JSON data. Builds up the necessary
+   * structure along the way, i.e., creates parent objects where necessary.
    *
-   * <p>The design suggestion for these methods is to have one per {@link
-   * services.question.ScalarType}
+   * @param path the {@link Path} with the fully specified path, e.g., "applicant.favorites.color"
+   *     or the equivalent "$.applicant.favorite.colors".
+   * @param value the value to place; values of type Map will create the equivalent JSON structure
    */
-  public Optional<String> readString(String path) {
+  public void put(Path path, Object value) {
+    path.parentPaths()
+        .forEach(
+            segmentPath -> {
+              if (this.jsonData.read(segmentPath.path()) == null) {
+                this.jsonData.put(
+                    segmentPath.parentPath().path(), segmentPath.keyName(), new HashMap<>());
+              }
+            });
+
+    this.jsonData.put(path.parentPath().path(), path.keyName(), value);
+  }
+
+  /** Same as the above, but accepts path as a string. */
+  public void put(String pathAsString, Object value) {
+    this.put(Path.create(pathAsString), value);
+  }
+
+  /**
+   * Returns the value at the given path, if it exists; otherwise returns {@link Optional#empty}.
+   *
+   * @param path the {@link Path} for the desired scalar
+   * @param type the expected type of the scalar
+   * @param <T> the expected type of the scalar
+   * @return optionally returns the value at the path if it exists, or empty if not
+   * @throws JsonPathTypeMismatchException if the scalar at that path is not the expected type
+   */
+  public <T> Optional<T> read(Path path, Class<T> type) throws JsonPathTypeMismatchException {
```

Comment:
Suggestion (blocking): similarly, please make this method private and instead expose public methods with more specific typing for the value to be read. It will allow us to customize how types are serialized for types that aren't supported by JSON.

Reply:
I added only readString for now and would prefer to add others when we have question types that require values other than String

Reply:
Sounds good!